docs: simplify installation page (#1751)
keptn/lifecycle-toolkit: docs/content/en/docs/getting-started.md

---

> title: Getting started with KLT

Reviewer: ```suggestion
title: Getting started with KLT Observability
```

Author: Small detail but this page gets users started with KLT (wholesale). Observability is one piece of KLT so I'm not going to change the titl.

- View DORA Metrics
- Install Grafana and Observability tooling to view DORA metrics

## System Overview

By the end of this page, here is what will be built.
This system will be built in stages.

![system overview](/docs/install/assets/install01.png)

## The Basics: A Deployment, Keptn and DORA Metrics

Let's start with the basics.
Here is what we will now build.

A deployment will occur.
Keptn will monitor the deployment and generate:

- An OpenTelemetry trace per deployment
- DORA metrics

![the basics](/docs/install/assets/install02.png)

Notice though that the metrics and traces have nowhere to go.
That will be fixed in a subsequent step.

## Step 1: Install Keptn Lifecycle Toolkit

Install the Keptn Lifecycle Toolkit using Helm:

```shell
helm repo add klt https://charts.lifecycle.keptn.sh
helm repo update
helm upgrade --install keptn klt/klt -n keptn-lifecycle-toolkit-system --create-namespace --wait
```

Keptn will need to know where to send OpenTelemetry traces.
Of course, Jaeger is not yet installed so traces have nowhere to go (yet),
but creating this configuration now means the system is preconfigured.

> Save this file as `collectorconfig.yaml`:

Reviewer: ```suggestion
Save this file as `collectorconfig.yaml`.
This connects your OpenTelemetry collector to KLT.
```

```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: keptndemo
  annotations:
    keptn.sh/lifecycle-toolkit: enabled
```

Create the namespace:

```shell
kubectl apply -f namespace.yaml
```

## Deploy Demo Application

It is time to deploy the demo application.

Save this manifest as `app.yaml`:

```shell
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  namespace: keptndemo
  labels:
    app.kubernetes.io/name: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/name: nginx
  template:
    metadata:
      labels:
        app.kubernetes.io/part-of: keptndemoapp
        app.kubernetes.io/name: nginx
        app.kubernetes.io/version: 0.0.1
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80
---
apiVersion: apps/v1
kind: Service
apiVersion: v1
kind: Service
metadata:
  name: nginx
  namespace: keptndemo
spec:
  selector:
    app.kubernetes.io/name: nginx
  ports:
    - protocol: TCP
      port: 8080
      targetPort: 80
```

Now apply:

```shell
kubectl apply -f app.yaml
```

> Keptn looks for these 3 labels:

Reviewer: ```suggestion
Keptn looks for these 3 annotations:
```

Author: I agree that we need to figure out the labels "vs." annotations thing but right now, it seems that labels is the preferred way. Leaving as-is

keptndemoapp-0.0.1-***    keptndemoapp   0.0.1     Completed
```

Keptn can run tasks and SLO evaluations before and after deployment.
You haven't configured this yet, but you can see the full lifecycle for a `keptnappversion` by running:

```shell
kubectl -n keptndemo get keptnappversion -o wide
```

Keptn applications are a collection of workloads.
By default, Keptn will build `KeptnApp` CRDs based on the labels you provide.

In the example above, the `KeptnApp` called `keptndemoapp` contains one workload (based on the `name` label):

## View your application

Port-forward to expose your app on `http://localhost:8080`:

```shell
kubectl -n keptndemo port-forward svc/nginx 8080
```

You should see the "Welcome to nginx" page.

![nginx demo app](/docs/install/assets/nginx.png)

## View DORA Metrics

Keptn is generating DORA metrics and OpenTelemetry traces for your deployments.

These metrics are exposed via the Keptn lifecycle operator `/metrics` endpoint on port `2222`.

To see these raw metrics, port-forward to the lifecycle operator metrics service:

```shell
kubectl -n keptn-lifecycle-toolkit-system port-forward service/keptn-klt-lifecycle-operator-metrics-service 2222

> ```

Reviewer: ```suggestion
```
Need to tell them that this is going to continue to run in their shell and that's okay
```

> To see these raw metrics, port-forward to the lifecycle operator metrics service:
> 
> ```shell
> kubectl -n keptn-lifecycle-toolkit-system port-forward service/keptn-klt-lifecycle-operator-metrics-service 2222
> ```
> 
> Access metrics in Prometheus format on `http://localhost:2222/metrics`.
> Look for metrics starting with `keptn_`.

Reviewer: ```suggestion
To see these raw metrics:

1. port-forward to the lifecycle operator metrics service:

   ```shell
   kubectl -n keptn-lifecycle-toolkit-system port-forward service/keptn-klt-lifecycle-       operator-metrics-service 2222
    ```

   Note that this will continue to run in your shell.
   If Docker is closed, you need to reissue this command
   to continue with this exercise
   or use this cluster to explore KLT more.

1. Access metrics in Prometheus format on `http://localhost:2222/metrics`.
    Look for metrics starting with `keptn_`.
```

![keptn prometheus metrics](/docs/install/assets/keptnprommetrics.png)

> 

Reviewer: ```suggestion

You need to open a new shell to continue.
```

## Make DORA metrics more user friendly

It is much more user friendly to provide dashboards for metrics, logs and traces.
So let's install new Observability components to help us:

- Cert manager: Jaeger requires cert-manager
- Jaeger: Store and view DORA deployment traces
- Prometheus: Store DORA metrics
- OpenTelemetry collector: Scrape metrics from the above DORA metrics endpoint & forward to Prometheus
- Grafana (and some prebuilt dashboards): Visualise the data

![add observability](/docs/install/assets/install01.png)

## Install Cert Manager

Jaeger requires Cert Manager, so install it now:

```shell
kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.12.2/cert-manager.crds.yaml
helm repo add jetstack https://charts.jetstack.io
helm repo update
helm install cert-manager --namespace cert-manager --version v1.12.2 jetstack/cert-manager --create-namespace --wait

> ```

Reviewer: ```suggestion
```
```
Is this the KLT cert-manager?  I thought that was now installed as part of the KLT installation.  Maybe only if you install from the manifest, which we did not do?  Need to clarify this with the core-engineering team.

Author: No it isn't. It is the "public" [default cert manager](https://cert-manager.io/). This is the root of my issue with "bundling" a custom cert manager. (The public / standard) cert manager is recommended for and indeed required by most tooling that communicates via TLS these days. So any realistic cluster configuration will end up with either:

1. KLT + it's cert manager
2. The default cert manager for everything else

OR

1. KLT w/o it's cert manager
2. The default cert manager for everything (inc. KLT)

Bundling a cert-manager with KLT seems to be a design decision taken solely to make demos easier.

---
title: Getting started with KLT
description: Get started with the Keptn Lifecycle Toolkit
weight: 10
hidechildren: false # this flag hides all sub-pages in the sidebar-multicard.html
---

Keptn Lifecycle Toolkit works whether or not you use a GitOps strategy.
The following is an imperative walkthrough.

## Prerequisites

- A Kubernetes cluster >= 1.24 (we recommend [Kubernetes kind](https://kind.sigs.k8s.io/docs/user/quick-start/))
  (`kind create cluster`)

> - [Helm](https://helm.io) CLI available

Reviewer: Also docker and kubectl

## Objectives

- Install Keptn Lifecycle Toolkit on your cluster
- Annotate a namespace and deployment to enable Keptn Lifecycle Toolkit
- View DORA Metrics
- Install Grafana and Observability tooling to view DORA metrics

## System Overview

By the end of this page, here is what will be built.
This system will be built in stages.

![system overview](/docs/install/assets/install01.png)

> ## The Basics: A Deployment, Keptn and DORA Metrics

Reviewer: Should this be entitled "Implement observability" or something?  Breaking this up into the three "use cases" might make it more digestible and also match the direction we're going with the product.

Author: I don't think so. That's fine for people familiar with the project or the "marketecture" of Keptn but it would mean little to a first time reader.

> Let's start with the basics.
> Here is what we will now build.
> 
> A deployment will occur.

Reviewer: ```suggestion
To begin our exploration of the Keptn observability features, we will:

* Deploy a simple application called `keptndemo`.
```

> Keptn will monitor the deployment and generate:

Reviewer: ```suggestion
* Keptn will monitor the deployment and generate:
```

> - An OpenTelemetry trace per deployment
> - DORA metrics

Reviewer: ```suggestion
   - An OpenTelemetry trace per deployment
   - DORA metrics
```

![the basics](/docs/install/assets/install02.png)

Notice though that the metrics and traces have nowhere to go.
That will be fixed in a subsequent step.

## Step 1: Install Keptn Lifecycle Toolkit

> Install Keptn Lifecycle Toolkit using Helm:

Reviewer: ```suggestion
Install the Keptn Lifecycle Toolkit using Helm:
```

> 

Reviewer: Need to first run `kind create cluster` here, right?

Author: Already in the top list where we mention kind

```shell
helm repo add klt https://charts.lifecycle.keptn.sh
helm repo update
helm upgrade --install keptn klt/klt -n keptn-lifecycle-toolkit-system --create-namespace --wait
```

Keptn will need to know where to send OpenTelemetry traces.
Of course, Jaeger is not yet installed so traces have nowhere to go (yet),
but creating this configuration now means the system is preconfigured.

> Save this file as `collectorconfig.yaml`:

Reviewer: Should we advise them where to put this?  Perhaps a suggestion that they create a directory on their workstation and put all these files there?  It seems like, in real life, they would subdivide that directory but maybe it's not necessary to introduce that complexity here.  And yes, they should know better than to just start creating these files in $HOME but...  And, for someone who has never explicitly worked with k8s before, it might not be obvious that you create all these files on your workstation.  Just saying -- a simple suggestion might be useful.

Author: I've added words to this effect.

```yaml
---
apiVersion: options.keptn.sh/v1alpha1
kind: KeptnConfig
metadata:
  name: keptnconfig-sample
  namespace: keptn-lifecycle-toolkit-system
spec:
  OTelCollectorUrl: 'jaeger-collector.keptn-lifecycle-toolkit-system.svc.cluster.local:4317'
  keptnAppCreationRequestTimeoutSeconds: 30
```

Apply the file and restart KLT to pick up the new config:

```shell
kubectl apply -f collectorconfig.yaml
kubectl rollout restart deployment -n keptn-lifecycle-toolkit-system -l control-plane=lifecycle-operator
kubectl rollout status deployment -n keptn-lifecycle-toolkit-system -l control-plane=lifecycle-operator --watch
kubectl rollout restart deployment -n keptn-lifecycle-toolkit-system -l component=scheduler
kubectl rollout status deployment -n keptn-lifecycle-toolkit-system -l component=scheduler --watch
```

## Create Namespace for Demo Application

Save this file as `namespace.yaml`.
The annotation means that Keptn Lifecycle Toolkit is active for workloads in this namespace.

```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: keptndemo

>   annotations:

Reviewer: It looks like this is one place where using `labels` is appropriate.

Reviewer 2: unfortunately we do not have labels check for the namespace enabled 😢

    keptn.sh/lifecycle-toolkit: enabled
```

Create the namespace:

```shell
kubectl apply -f namespace.yaml
```

## Deploy Demo Application

It is time to deploy the demo application.

Save this manifest as `app.yaml`:

```shell
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  namespace: keptndemo
  labels:
    app.kubernetes.io/name: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/name: nginx
  template:
    metadata:

>       labels:

Reviewer: Should we instead use `annotations` here?

        app.kubernetes.io/part-of: keptndemoapp
        app.kubernetes.io/name: nginx
        app.kubernetes.io/version: 0.0.1
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80
---
apiVersion: apps/v1
kind: Service
apiVersion: v1
kind: Service
metadata:
  name: nginx
  namespace: keptndemo
spec:
  selector:
    app.kubernetes.io/name: nginx
  ports:
    - protocol: TCP
      port: 8080
      targetPort: 80
```

Now apply:

```shell
kubectl apply -f app.yaml
```

Keptn looks for these 3 labels:

- `app.kubernetes.io/part-of`
- `app.kubernetes.io/name`
- `app.kubernetes.io/version`

These are [Kubernetes recommended labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels)
but if you want to use different labels, you can swap for:

- `keptn.sh/app` instead of `app.kubernetes.io/part-of`
- `keptn.sh/workload` instead of `app.kubernetes.io/name`

> - `keptn.sh/version` instead of `app.kubernetes.io/version`

Reviewer: Is this necessary here?  We aren't providing much context to anything in this exercise -- why is this tidbit worthy of mention?  If anything, I'd provide a link to the main docs that explains all this.  I would probably include a lot more links to the official docs in this piece but then I probably tend to over-complicate things. Dunno.

Reviewer 2: IMHO this is important for first users and should stay. It is very nice that one can use standard k8s labels, i.e. if I already have them in my 100 deployments manifests, I do not need to do anything at all to use klt for them except enable their namespace.

## Explore Keptn

Keptn is now aware of your deployments and is generating DORA statistics about them.

> Keptn has created a CRD to track your application.

Reviewer: ```suggestion
Keptn has created a resource to track your application.
```
Flo corrected me about terminology:  A CRD is the definition itself; in other words, the syntax.  Each instance that is created using that CRD is just a CR or resource.  All the Keptn specific resources are CRs but the pure Kubernetes resources are not "Custom" so just resources.  So I'm just calling them all resources.

> The name of which is based on the `part-of` label.

Reviewer: This would become "annotation" if you modify the recommended content of that file.

It may take up to 30 seconds to create the `KeptnApp` so run the following command until you see the `keptnappdemo` CRD.

```shell
kubectl -n keptndemo get keptnapp
```

Expected output:

```shell
NAME           AGE
keptndemoapp   2s
```

Keptn also creates a new application version every time you increment the `version` label..

The `PHASE` will change as the deployment progresses.
A successful deployment is shown as `PHASE=Completed`

```shell
kubectl -n keptndemo get keptnappversion
```

Expected output:

```shell
NAME                      APPNAME        VERSION   PHASE
keptndemoapp-0.0.1-***    keptndemoapp   0.0.1     Completed
```

Keptn can run tasks and SLO evaluations before and after deployment.
You haven't configured this yet, but you can see the full lifecycle for a `keptnappversion` by running:

```shell
kubectl -n keptndemo get keptnappversion -o wide
```

Keptn applications are a collection of workloads.

> By default, Keptn will build `KeptnApp` CRDs based on the labels you provide.

Reviewer: ```suggestion
By default, Keptn builds `KeptnApp` resources based on the annotations and labels you provide to the workloads.
```
Is this a place to list the workloads to which you can apply the annotations/labels?  I'm thinking maybe not -- that is for the next stage of their education when they read the main docs, right?

Author: I agree, this is the "get me going quick - I don't need to know" file.

> In the example above, the `KeptnApp` called `keptndemoapp` contains one workload (based on the `name` label):

Reviewer: How hard would it be to include two workloads in this demo?  That just seems like a good way to emphasize that one KeptnApp can include multiple workloads.

Author: We could easily switch out the demo to use the OpenTelemetry demo (which seems to be moving towards the industry standard demo app) BUT it requires 4GB just for the demo app.

We could also use [podtato head](https://github.com/podtato-head/podtato-head) as that's the standard from the TAG App Delivery folk.

## View your application

Port-forward to expose your app on `http://localhost:8080`:

```shell
kubectl -n keptndemo port-forward svc/nginx 8080
```

You should see the "Welcome to nginx" page.

![nginx demo app](/docs/install/assets/nginx.png)

## View DORA Metrics

> Keptn is generating DORA metrics and OpenTelemetry traces for your deployments.

Reviewer: Where did we configure the oTel collector and create the KeptnConfig file that shows its location?

```shell
NAME                      APPNAME        VERSION   PHASE
keptndemoapp-0.0.1-***    keptndemoapp   0.0.1     Completed
```

Keptn can run tasks and SLO evaluations before and after deployment.
You haven't configured this yet, but you can see the full lifecycle for a `keptnappversion` by running:

```shell
kubectl -n keptndemo get keptnappversion -o wide
```

Keptn applications are a collection of workloads.
By default, Keptn will build `KeptnApp` CRDs based on the labels you provide.

In the example above, the `KeptnApp` called `keptndemoapp` contains one workload (based on the `name` label):

## View your application

Port-forward to expose your app on `http://localhost:8080`:

```shell
kubectl -n keptndemo port-forward svc/nginx 8080
```

Open a browser window and go to `http://localhost:8080`

You should see the "Welcome to nginx" page.

![nginx demo app](../assets/nginx.png)

## View DORA Metrics

Keptn is generating DORA metrics and OpenTelemetry traces for your deployments.

These metrics are exposed via the Keptn lifecycle operator `/metrics` endpoint on port `2222`

To see these raw metrics:

- Port forward to the lifecycle operator metrics service:

```shell

> kubectl -n keptn-lifecycle-toolkit-system port-forward service/keptn-klt-lifecycle-operator-metrics-service 2222

Reviewer 2: of course, this adds jq as a prerequisite, but maybe saves us refactoring later? 🤔

> kubectl -n keptn-lifecycle-toolkit-system port-forward service/keptn-klt-lifecycle-operator-metrics-service 2222

Reviewer 3: to avoid `jq` we can use
```suggestion
SERVICE=$(kubectl get svc -l control-plane=lifecycle-operator -A -ojsonpath="{.items[0].metadata.name}")
kubectl -n keptn-lifecycle-toolkit-system port-forward svc/$SERVICE 2222
```
or even more compress
```suggestion
kubectl -n keptn-lifecycle-toolkit-system port-forward svc/$(kubectl get svc -l control-plane=lifecycle-operator -A -ojsonpath="{.items[0].metadata.name}") 2222```

```

Note that this command will (and should) continue to run in your terminal windows.
Open a new terminal window to continue.

- Access metrics in Prometheus format on `http://localhost:2222/metrics`
- Look for metrics starting with `keptn_`.

![keptn prometheus metrics](../assets/keptnprommetrics.png)

Keptn emits various metrics about the state of your system.
These metrics can then be visualised in Grafana.

For example:

- `keptn_app_active` tracks the number of applications that Keptn manages
- `keptn_deployment_active` tracks the currently live number of deployments occuring.

>   Expect this metrics to be `0` when everything is currently deployed.

Reviewer 4: ```suggestion
  Expect this metric to be `0` when everything is currently deployed.
```

  It will occasionally rise to `n` during deployments and then fall back to `0` when deployments are completed.

There are many other Keptn metrics.

## Make DORA metrics more user friendly

It is much more user friendly to provide dashboards for metrics, logs and traces.
So let's install new Observability components to help us:

- Cert manager: Jaeger requires cert-manager
- Jaeger: Store and view DORA deployment traces
- Prometheus: Store DORA metrics
- OpenTelemetry collector: Scrape metrics from the above DORA metrics endpoint & forward to Prometheus

> - Grafana (and some prebuilt dashboards): Visualise the data

Reviewer 4: [nitpick] I suggest adding links to the various projects

kubectl apply -f jaeger.yaml -n keptn-lifecycle-toolkit-system
kubectl wait --for=condition=available deployment/jaeger -n keptn-lifecycle-toolkit-system --timeout=300s
```

Port-forward to access Jaeger:

```shell
kubectl -n keptn-lifecycle-toolkit-system port-forward svc/jaeger-query 16686
```

Jaeger is available on `http://localhost:16686`

## Install Grafana dashboards

Create some Keptn Grafana dashboards that will be available when Grafana is installed and started:

```shell
kubectl create ns monitoring
kubectl apply -f https://raw.githubusercontent.com/keptn/lifecycle-toolkit/main/examples/support/observability/config/prometheus/grafana-config.yaml
kubectl apply -f https://raw.githubusercontent.com/keptn/lifecycle-toolkit/main/examples/support/observability/config/prometheus/grafana-dashboard-keptn-applications.yaml
kubectl apply -f https://raw.githubusercontent.com/keptn/lifecycle-toolkit/main/examples/support/observability/config/prometheus/grafana-dashboard-keptn-overview.yaml
kubectl apply -f https://raw.githubusercontent.com/keptn/lifecycle-toolkit/main/examples/support/observability/config/prometheus/grafana-dashboard-keptn-workloads.yaml
```

> ## Install Grafana datasources

Reviewer 4: Those links should point to the latest released versions of the files and not to main. Then they can be updated with release-please during release of KLT.
Something like: (please look at the raw code of this comment, the release please annotation comments are hidden in the preview)

<!---x-release-please-start-version-->
```shell
kubectl create ns monitoring
kubectl apply -f https://raw.githubusercontent.com/keptn/lifecycle-toolkit/klt-v0.8.1/examples/support/observability/config/prometheus/grafana-config.yaml
kubectl apply -f https://raw.githubusercontent.com/keptn/lifecycle-toolkit/klt-v0.8.1/examples/support/observability/config/prometheus/grafana-dashboard-keptn-applications.yaml
kubectl apply -f https://raw.githubusercontent.com/keptn/lifecycle-toolkit/klt-v0.8.1/examples/support/observability/config/prometheus/grafana-dashboard-keptn-overview.yaml
kubectl apply -f https://raw.githubusercontent.com/keptn/lifecycle-toolkit/klt-v0.8.1/examples/support/observability/config/prometheus/grafana-dashboard-keptn-workloads.yaml
```
<!---x-release-please-end-->

                "version": 1
            },
            {
                "orgId":1,
                "name":"Jaeger",
                "type":"jaeger",
                "typeName":"Jaeger",
                "typeLogoUrl":"public/app/plugins/datasource/jaeger/img/jaeger_logo.svg",
                "access":"proxy",
                "url":"http://jaeger-query.keptn-lifecycle-toolkit-system.svc.cluster.local:16686",
                "user":"",
                "database":"",
                "basicAuth":false,
                "isDefault":false,
                "jsonData":{"spanBar":{"type":"None"}},
                "readOnly":false
            }
        ]
    }
```

Now apply it:

```shell
kubectl apply -f datasources.yaml
```

## Install kube prometheus stack

This will install:

- Prometheus
- Prometheus Configuration
- Grafana & default dashboards

Save this file as `values.yaml`:

```yaml
grafana:
  adminPassword: admin
  sidecar.datasources.defaultDatasourceEnabled: false
prometheus:
  prometheusSpec:
    additionalScrapeConfigs:
      - job_name: "scrape_klt"
        scrape_interval: 5s
        static_configs:

>           - targets: ['keptn-klt-lifecycle-operator-metrics-service.keptn-lifecycle-toolkit-system.svc.cluster.local:2222']

Reviewer 2: again we could use $SERVICE instead

```

```shell
helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
helm repo update
helm upgrade --install observability-stack prometheus-community/kube-prometheus-stack --version 48.1.1 --namespace monitoring --values=values.yaml --wait
```

## Access Grafana

```shell
kubectl -n monitoring port-forward svc/observability-stack-grafana 80
```

- Grafana username: `admin`
- Grafana password: `admin`

View the Keptn dashboards at: `http://localhost/dashboards`

Remember that Jaeger and Grafana weren't installed during the first deployment
so expect the dashboards to look a little empty.

## Deploy v0.0.2 and populate Grafana

By triggering a new deployment, Keptn will track this deployment and the Grafana dashboards will actually have data.

> Modify your `app.yaml` and change the `app.kubernetes.io/version` from `0.0.1` to `0.0.2`.

Reviewer 4: ```suggestion
Modify your `app.yaml` and change the `app.kubernetes.io/version` from `0.0.1` to `0.0.2` (or `keptn.sh/version` if you used the Keptn specific labels earlier).
```

Apply your update:

```shell
kubectl apply -f app.yaml
```

After about 30 seconds you should now see two `keptnappversions`:

```shell
kubectl -n keptndemo get keptnappversion
```

Expected output:

```shell
NAME                          APPNAME        VERSION   PHASE

> keptndemoapp-0.0.1-6b86b273   keptndemoapp   0.0.1     Completed
> keptndemoapp-0.0.2-d4735e3a   keptndemoapp   0.0.2     AppDeploy

Reviewer 4: ```suggestion
keptndemoapp-0.0.1-***  keptndemoapp   0.0.1     Completed
keptndemoapp-0.0.2-***  keptndemoapp   0.0.2     AppDeploy
```

```

Wait until the `PHASE` of `keptndemoapp-0.0.2` is `Completed`.
This signals that the deployment was successful and the pod is running.

View the Keptn Applications Dashboard and you should see the DORA metrics and an OpenTelemetry per trace:

![keptn applications dashboard](../assets/keptnapplications.png)

![deployment trace](../assets/deploymenttrace.png)

## More control over KeptnApp

You may have noticed that the `KeptnApp` Custom Resources are created automatically by KLT.

> The lifecycle toolkit automatically groups workloads into `KeptnApp` by looking for matching `part-of` annotations.

Reviewer 4: ```suggestion
The lifecycle toolkit automatically groups workloads into `KeptnApp`s by looking for matching `part-of` annotations.
```

> Any workloads with the same `part-of` annotation is said to be `part-of` the same `KeptnApp`.

Reviewer 4: ```suggestion
Any workloads with the same `part-of` annotation are said to be `part-of` the same `KeptnApp`.
```

However, you can override this automatic behaviour by creating a custom `KeptnApp` CRD.
In this way, you are in full control of what constitutes a Keptn Application.

> See [Define a Keptn Application](../implementing/integrate/#define-keptnapp-manually) for more information.

Reviewer 4: this link here seems to be broken in the deploy preview... i don't know why. maybe the path is not right?

## What's next?

Keptn can run pre and post deployment tasks and SLO evaluations automatically.

> Continue to Keptn learning journey by [adding deployment tasks](https://example.com).

Reviewer 4: ```suggestion
Continue to Keptn learning journey by [adding deployment tasks](../tutorials/add-tasks/).
```

- Deploy a simple application called `keptndemo`.

Keptn will monitor the deployment and generate:

- An OpenTelemetry trace per deployment
- DORA metrics

![the basics](/docs/assets/install02.png)

Notice though that the metrics and traces have nowhere to go.
That will be fixed in a subsequent step.

## Step 1: Install Keptn Lifecycle Toolkit

Install the Keptn Lifecycle Toolkit using Helm:

```shell
helm repo add klt https://charts.lifecycle.keptn.sh
helm repo update
helm upgrade --install keptn klt/klt -n keptn-lifecycle-toolkit-system --create-namespace --wait
```

Keptn will need to know where to send OpenTelemetry traces.
Of course, Jaeger is not yet installed so traces have nowhere to go (yet),
but creating this configuration now means the system is preconfigured.

Save this file as `collectorconfig.yaml`.
It doesn't matter where this file is located on your local machine:

```yaml
---
apiVersion: options.keptn.sh/v1alpha1
kind: KeptnConfig
metadata:
  name: keptnconfig-sample
  namespace: keptn-lifecycle-toolkit-system
spec:
  OTelCollectorUrl: 'jaeger-collector.keptn-lifecycle-toolkit-system.svc.cluster.local:4317'
  keptnAppCreationRequestTimeoutSeconds: 30
```

> Apply the file and restart KLT to pick up the new config:

Reviewer 5: @agardnerIT did you also try to apply the `keptnconfig-sample` without restarting the deployments? The restart should not be required, and the new config should be applied on the fly within a couple of seconds. If that is not the case we need to create a bug issue and investigate this further

Reviewer 4: i agree, let's move this to a follow-up

```shell
kubectl apply -f collectorconfig.yaml
kubectl rollout restart deployment -n keptn-lifecycle-toolkit-system -l control-plane=lifecycle-operator
kubectl rollout status deployment -n keptn-lifecycle-toolkit-system -l control-plane=lifecycle-operator --watch
kubectl rollout restart deployment -n keptn-lifecycle-toolkit-system -l component=scheduler
kubectl rollout status deployment -n keptn-lifecycle-toolkit-system -l component=scheduler --watch
```

## Step 2: Create Namespace for Demo Application

Save this file as `namespace.yaml`.
The annotation means that Keptn Lifecycle Toolkit is active for workloads in this namespace.

```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: keptndemo
  annotations:
    keptn.sh/lifecycle-toolkit: enabled
```

Create the namespace:

```shell
kubectl apply -f namespace.yaml
```

## Step 3: Deploy Demo Application

It is time to deploy the demo application.

Save this manifest as `app.yaml`:

> ```shell

Reviewer 5: ```suggestion
```yaml
```

but if you want to use different labels, you can swap them for these KLT specific labels:

- `keptn.sh/app` instead of `app.kubernetes.io/part-of`
- `keptn.sh/workload` instead of `app.kubernetes.io/name`
- `keptn.sh/version` instead of `app.kubernetes.io/version`

## Step 4: Explore Keptn

Keptn is now aware of your deployments and is generating DORA statistics about them.

Keptn has created a resource called a `KeptnApp` to track your application.
The name of which is based on the `part-of` label.

It may take up to 30 seconds to create the `KeptnApp` so run the following command until you see the `keptnappdemo` CRD.

```shell
kubectl -n keptndemo get keptnapp
```

Expected output:

```shell
NAME           AGE
keptndemoapp   2s
```

Keptn also creates a new application version every time you increment the `version` label.

The `PHASE` will change as the deployment progresses.
A successful deployment is shown as `PHASE=Completed`

```shell
kubectl -n keptndemo get keptnappversion
```

Expected output:

```shell
NAME                      APPNAME        VERSION   PHASE
keptndemoapp-0.0.1-***    keptndemoapp   0.0.1     Completed
```

> Keptn can run tasks and SLO evaluations before and after deployment.

Reviewer 5: I think the term SLO has not been mentioned at this point yet, so we might spell it out here
```suggestion
Keptn can run tasks and SLO (Service Level Objectives) evaluations before and after deployment.
```

You haven't configured this yet, but you can see the full lifecycle for a `keptnappversion` by running:

```shell
kubectl -n keptndemo get keptnappversion -o wide
```

Keptn applications are a collection of workloads.
By default, Keptn will build `KeptnApp` resources based on the labels you provide.

> In the example above, the `KeptnApp` called `keptndemoapp` contains one `KeptnWorkload` (based on the `name` label):

Reviewer 5: ```suggestion
In the example above, the `KeptnApp` called `keptndemoapp` contains one `KeptnWorkload` (based on the `app.kubernetes.io/name` label):
```

## Step 7: Make DORA metrics more user friendly

It is much more user friendly to provide dashboards for metrics, logs and traces.
So let's install new Observability components to help us:

- [Cert manager](https://cert-manager.io): Jaeger requires cert-manager
- [Jaeger](https://jaegertracing.io): Store and view DORA deployment traces
- [Prometheus](https://prometheus.io): Store DORA metrics
- [OpenTelemetry collector](https://opentelemetry.io/docs/collector/):
  Scrape metrics from the above DORA metrics endpoint & forward to Prometheus
- [Grafana](https://grafana.com) (and some prebuilt dashboards): Visualise the data

![add observability](/docs/assets/install01.png)

## Step 8: Install Cert Manager

Jaeger requires Cert Manager, so install it now:

```shell
kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.12.2/cert-manager.crds.yaml
helm repo add jetstack https://charts.jetstack.io
helm repo update
helm install cert-manager --namespace cert-manager --version v1.12.2 jetstack/cert-manager --create-namespace --wait
```

## Step 9: Install Jaeger

Save this file as `jaeger.yaml` (it can be saved anywhere on your computer):

> ```shell

Reviewer 5: ```suggestion
```yaml
```

```shell
kubectl -n monitoring port-forward svc/observability-stack-grafana 80
```

- Grafana username: `admin`
- Grafana password: `admin`

View the Keptn dashboards at: `http://localhost/dashboards`

Remember that Jaeger and Grafana weren't installed during the first deployment
so expect the dashboards to look a little empty.

## Step 13: Deploy v0.0.2 and populate Grafana

By triggering a new deployment, Keptn will track this deployment and the Grafana dashboards will actually have data.

Modify your `app.yaml` and change the `app.kubernetes.io/version` from `0.0.1` to `0.0.2`
(or `keptn.sh/version` if you used the Keptn specific labels earlier).

Apply your update:

```shell
kubectl apply -f app.yaml
```

After about 30 seconds you should now see two `keptnappversions`:

```shell
kubectl -n keptndemo get keptnappversion
```

Expected output:

```shell
NAME                          APPNAME        VERSION   PHASE
keptndemoapp-0.0.1-***  keptndemoapp   0.0.1     Completed
keptndemoapp-0.0.2-***  keptndemoapp   0.0.2     AppDeploy
```

Wait until the `PHASE` of `keptndemoapp-0.0.2` is `Completed`.
This signals that the deployment was successful and the pod is running.

> View the Keptn Applications Dashboard and you should see the DORA metrics and an OpenTelemetry per trace:

Reviewer 5: ```suggestion
View the Keptn Applications Dashboard and you should see the DORA metrics and an OpenTelemetry trace:
```
